addons/mail/models/stage_time_mixin.py

```diff
@@ -0,0 +1,111 @@
+from odoo import api, fields, models
+
+
+class StageTimeMixin(models.AbstractModel):
+    _name = "stage.time.mixin"
+    _description = "Mixin to compute the time a record spent in each stage"
+    _stage_field = "stage_id"
+
+    time_per_stage = fields.Char(compute="_compute_time_per_stage",
+                                 help="JSON string that maps stage ids to ms and days",
+                                 store=True)
+
+    @api.depends(lambda self: [self._stage_field])
+    def _compute_time_per_stage(self):
+        for record in self:
+            if isinstance(record.id, models.NewId):
+                record.time_per_stage = record.time_per_stage
+            else:
+                record.time_per_stage = record._get_json_ms_in_stages()
+
+    def _get_json_ms_in_stages(self):
+        """
+        :return: string: JSON {stage_id (str): ms (float)}
+
+        e.g.:    '{"3": 1000, "2": 44211.3331}'
+        """
+
+        stages_with_time_json = "{}"
+        # Should not compute if the record was just created or created without stage and then assigned a stage
+        if (self.env.cr.precommit.data
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}']
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id] is not None
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field]
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field] != self.stage_id):
```

Comment:
Ah, actually this won't work since it's `!=`.

Still, I don't think it would hurt to at least extract the `self.env.cr.precommit.data and self.env.cr.precommit.data.get(f'mail.tracking.{self._name}')` to keep the lines short and avoid noise